src/states/approved_peer/tests/elder.rs

```diff
@@ -0,0 +1,598 @@
+        self.other_ids = new_info.new_other_ids;
+        self.accumulate_self_ack_message();
+    }
+}
+
+fn create_state(
+    rng: &mut MainRng,
+    network: &Network,
+    full_id: FullId,
+    secret_key_share: bls::SecretKeyShare,
+    gen_pfx_info: GenesisPfxInfo,
+) -> ApprovedPeer {
+    let parsec_map = ParsecMap::default().with_init(rng, full_id.clone(), &gen_pfx_info);
+    let chain = Chain::new(
+        Default::default(),
+        *full_id.public_id(),
+        gen_pfx_info.clone(),
+        Some(secret_key_share),
+    );
+
+    let details = ElderDetails {
+        chain,
+        network_service: test_utils::create_network_service(network),
+        event_backlog: Default::default(),
+        full_id,
+        gen_pfx_info,
+        msg_queue: Default::default(),
+        sig_accumulator: Default::default(),
+        parsec_map,
+        msg_filter: MessageFilter::new(),
+        timer: test_utils::create_timer(),
+        rng: rng::new_from(rng),
+    };
+
+    let elder = ApprovedPeer::from_joining_peer(details, Connected::First, &mut ());
+    assert!(elder.chain.is_self_elder());
+    elder
+}
+
+fn gen_p2p_node(rng: &mut MainRng, network: &Network) -> P2pNode {
+    P2pNode::new(*FullId::gen(rng).public_id(), network.gen_addr())
+}
+
+#[test]
+fn construct() {
+    let env = Env::new(ELDER_SIZE - 1);
+
+    assert!(!env.has_unpolled_observations());
+    assert!(!env.is_candidate_elder());
+}
+
+#[test]
+fn when_accumulate_online_then_node_is_added_to_our_members() {
+    let mut env = Env::new(ELDER_SIZE - 1);
+    let new_info = env.new_elders_info_with_candidate();
+    env.accumulate_online(env.candidate.clone());
+    env.accumulate_start_dkg(&new_info);
+
+    assert!(!env.has_unpolled_observations());
+    assert!(env.is_candidate_member());
+    assert!(!env.is_candidate_elder());
+}
+
+#[test]
+fn when_accumulate_online_and_start_dkg_and_section_info_then_node_is_added_to_our_elders() {
+    let mut env = Env::new(ELDER_SIZE - 1);
+    let new_info = env.new_elders_info_with_candidate();
+    env.accumulate_online(env.candidate.clone());
+    env.accumulate_start_dkg(&new_info);
+
+    env.accumulate_section_info_if_vote(&new_info);
+    env.accumulate_voted_unconsensused_events();
+
+    assert!(!env.has_unpolled_observations());
+    assert!(env.is_candidate_member());
+    assert!(env.is_candidate_elder());
+}
+
+#[test]
+fn when_accumulate_offline_then_node_is_removed_from_our_members() {
+    let mut env = Env::new(ELDER_SIZE - 1);
+    let new_info = env.new_elders_info_with_candidate();
+    env.accumulate_online(env.candidate.clone());
+    env.accumulate_start_dkg(&new_info);
+    env.accumulate_section_info_if_vote(&new_info);
+    env.accumulate_voted_unconsensused_events();
+
+    env.other_ids = new_info.new_other_ids;
+    let new_info = env.new_elders_info_without_candidate();
+    env.accumulate_offline(*env.candidate.public_id());
+    env.accumulate_start_dkg(&new_info);
```

Comment:
what if not accumulate start_dkg, just accumulate offline ?

Reply:
Not sure what you are asking. That we should have a test for that? If so, then yeah, maybe. But again, this would be out of scope of this PR.

Reply:
it's just not clear to me that whether it is `accumulate_offline` or `accumulate_start_dkg` cause the node to be removed from members but still retain as an elder.
especially the next test shows that to be removed fron elders, there are extra steps after `accumulate_start_dkg`. (I thought)

Reply:
No, I think it's accumulate Offline first, then accumulate DKG and then accumulate SectionInfo. Only then is the node no longer considered elder. But in any case, the logic of these tests is still the same as in fleming (only some type/variable names changed), so I don't think it's appropriate to change it here.